xpart/particles/particles.py

```diff
@@ -157,109 +159,346 @@ class Particles(xo.HybridClass):
+        if mask is None:
+            setattr(self, varname, target_val)
+            return
+
+        # Assign with a mask
+        if isinstance(self._context, xo.ContextPyopencl):  # PyOpenCL array
+            if hasattr(mask, 'get'):
+                mask = mask.get()
+            mask = np.where(mask)[0]
+            mask = self._context.nparray_to_context_array(mask)
+
+        getattr(self, varname)[mask] = target_val[mask]
+
+    def _update_refs(self, p0c=None, energy0=None, gamma0=None, beta0=None,
+                     mask=None):
+        if not any(ff is not None for ff in (p0c, energy0, gamma0, beta0)):
+            self._p0c = 1e9
+            p0c = self._p0c
+
+        if p0c is not None:
+            _energy0 = self._sqrt(p0c ** 2 + self.mass0 ** 2)
+            _beta0 = p0c / _energy0
+            _gamma0 = _energy0 / self.mass0
+            _p0c = p0c
+        elif energy0 is not None:
+            _p0c = self._sqrt(energy0 ** 2 - self.mass0 ** 2)
+            _beta0 = _p0c / energy0
+            _gamma0 = energy0 / self.mass0
+        elif gamma0 is not None:
+            _beta0 = self._sqrt(1 - 1 / gamma0 ** 2)
+            _energy0 = self.mass0 * gamma0
+            _p0c = _energy0 * _beta0
+            _gamma0 = gamma0
+        elif beta0 is not None:
+            _gamma0 = 1 / self._sqrt(1 - beta0 ** 2)
+            _energy0 = self.mass0 * _gamma0
+            _p0c = _energy0 * beta0
+            _beta0 = beta0
+        else:
+            raise RuntimeError('This statement is unreachable.')
+
+        self._assert_values_consistent(energy0, self.mass0 * _gamma0, mask)
+        self._setattr_if_consistent('_p0c', p0c, _p0c, mask)
+        self._setattr_if_consistent('_gamma0', gamma0, _gamma0, mask)
+        self._setattr_if_consistent('_beta0', beta0, _beta0, mask)
+
+    def _update_energy_deviations(self, delta=None, ptau=None, pzeta=None,
+                                  _rpp=None, _rvv=None, mask=None):
+        if all(ff is None for ff in (delta, ptau, pzeta)):
+            if _rpp is not None or _rvv is not None:
+                raise ValueError('Setting `delta` and `ptau` by only giving '
+                                 '`_rpp` and `_rvv` is not supported.')
+            self._delta = 0.0
+            delta = self._delta  # Cupy complains if we later assign LinkedArray
+
+        beta0 = self._beta0
+        if delta is not None:
+            _delta = delta
+            _ptau = self._sqrt(_delta**2 + 2 * _delta + 1 / beta0**2) - 1 / beta0
+            _pzeta = _ptau / beta0
+        elif ptau is not None:
+            _ptau = ptau
+            _delta = self._sqrt(_ptau ** 2 + 2 * _ptau / beta0 + 1) - 1
+            _pzeta = _ptau / beta0
+        elif pzeta is not None:
+            _pzeta = pzeta
+            _ptau = _pzeta * beta0
+            _delta = self._sqrt(_ptau ** 2 + 2 * _ptau / beta0 + 1) - 1
+        else:
+            raise RuntimeError('This statement is unreachable.')
```

Comment:
Check consistency with the physics guide:
https://github.com/xsuite/xsuite/blob/main/docs/physics_manual/physics_man.pdf